File: graph/BFS/python/BFS.py
from collections import deque
from collections.abc import Sequence


def BFS(graph: Sequence[Sequence[int]], start: int) -> list[int]:
    order: list[int] = []
    visited = [False] * len(graph)
    queue: deque[int] = deque([start])
    visited[start] = True

    while queue:
        node = queue.popleft()
        order.append(node)

        for neighbor in graph[node]:
            if not visited[neighbor]:
                visited[neighbor] = True
                queue.append(neighbor)

    return order

File: graph/BFS/rust/src/BFS.rs
use std::collections::VecDeque;

#[allow(non_snake_case)]
pub fn BFS(graph: &[Vec<usize>], start: usize) -> Vec<usize> {
    let mut order = Vec::new();
    let mut visited = vec![false; graph.len()];
    let mut queue = VecDeque::from([start]);

    visited[start] = true;

    while let Some(node) = queue.pop_front() {
        order.push(node);

        for &neighbor in &graph[node] {
            if !visited[neighbor] {
                visited[neighbor] = true;
                queue.push_back(neighbor);
            }
        }
    }

    order
}
